tools/pre_ci.py

```
@@ -0,0 +1,201 @@
+
+    def record_result(self, description: str, passed: bool | str) -> None:
+        """Records the outcome of a specific step for the final summary."""
+        self._results.append((description, passed))
+
+    def run_command(
+        self, command: list[str], description: str, fail_fast: bool = False
+    ) -> bool:
+        """Executes a single shell command and prints the status."""
+        print(f"\n>>> [Step: {description}]")
+        try:
+            subprocess.run(command, check=True)
```

Ruff (0.15.11)

[error] `subprocess` call: check for execution of untrusted input

(S603)